### DISCUSSION TITLE
How to write (update) immutable classes

### DISCUSSION THREAD

**Discussion by wpl90** [isar/isar#240] [Help] [answered]
Hello, thank you for developing a great package called Isar.

I would like to ask you one question.

https://isar.dev/schema.html#annotating-classes

@Collection()
class Contact {
 @Id()
 int? id;

 String firstName;

 String lastName;

 bool isStarred;
}

In the documentation, as shown above, all the fields of the model class are declared as mutable. Is it the only way to declare Isar's model class as mutable?

I'm using Isar in a model class with immutable fields, but I'm not sure how to write to the isar collection.

`  @override
  Future<void> updateRecentDateTimeOfPractice(int index, dynamic tempQDSs) async {
    await isar.writeTxn((isar) async {
      Word? target = await isar.words.get((tempQDSs as List<Word>)[index].id!);
      Word newWord = Word(
        body:target!.body,
        rank:target!.rank,
        numOfCorrect: target!.numOfCorrect,
        numOfWrong: target!.numOfWrong-1,
        createdAt: target!.createdAt,
        recentDateTimeOfPractice: DateTime.now(),
        memo: target!.memo,
      );

      isar.words.put(newWord);
    }).then((_){
      print('successsssss');
    }).catchError((e){
      print(e.toString());
    });;
  }`

I want to update existing items with the above code, but for some reason new items keep getting added to the collection.

Is there a concrete example of writing (updating) an immutable model class somewhere?

**Comment by simc** [comment, marked as answer]
The difference with immutable classes is that Isar cannot update the id of your object after you put it (because it is immutable).

Let's say you have this class:
```dart
@Collection()
class Contact {
  @id()
  final int? id;
  
  final String firstName;
  
  Contact(this.id, this.firstName);
  
  Contact copyWith(int id) {
    return Contact(id, name);
  }
}
```

You need to save the object like this:
```dart
var contact = Contact(null, 'David');
isar.writeTxn((isar) async {
  final newId = await isar.contacts.put(contact);
  contact = contact.copyWith(id);
});
```

**Reply by wpl90** [reply]
Thanks for the reply.

I'm sorry, I don't fully understand, don't you use "newId" in the sample code you showed?

Just to confirm, the code you showed me is the code to add a new contact named David to the collection, right?

When you say we can't change the id, do you mean we can't change it because it's declared using "final"?

If that's the case, "firstName" is also declared using "final", can't we(users of this app) change "firstName" later?

If I declare the model class immutable in that way, can the user of the app change the name of the registered contact (David) to another name, for example, with the code you showed?

If we can, can you tell me how exactly I can write the code to implement this feature?
